src/main/java/org/rumbledb/compiler/VariableDependenciesVisitor.java

```diff
@@ -0,0 +1,466 @@
+        }
+        this.inputVariableDependencies.get(node).addAll(variables);
+    }
+
+    private void removeInputVariableDependencies(Node node, Set<String> variables) {
+        if (variables == null) {
+            addInputVariableDependencies(node, Collections.emptySet());
+            return;
+        }
+        if (!this.inputVariableDependencies.keySet().contains(node)) {
+            return;
+        }
+        for (String v : variables) {
+            this.inputVariableDependencies.get(node).remove(v);
+        }
+    }
+
+    private void addInputVariableDependency(Node node, String variable) {
+        if (variable == null) {
+            addInputVariableDependencies(node, Collections.emptySet());
+            return;
+        }
+        if (!this.inputVariableDependencies.keySet().contains(node)) {
+            this.inputVariableDependencies.put(node, new TreeSet<String>());
+        }
+        this.inputVariableDependencies.get(node).add(variable);
+    }
+
+    private void removeInputVariableDependency(Node node, String variable) {
+        if (variable == null) {
+            addInputVariableDependencies(node, Collections.emptySet());
+            return;
+        }
+        if (!this.inputVariableDependencies.keySet().contains(node)) {
+            return;
+        }
+        this.inputVariableDependencies.get(node).remove(variable);
+    }
+
+    private void addOutputVariableDependencies(Node node, Set<String> variables) {
+        if (variables == null) {
+            addInputVariableDependencies(node, Collections.emptySet());
+            return;
+        }
+        if (!this.outputVariableDependencies.keySet().contains(node)) {
+            this.outputVariableDependencies.put(node, new TreeSet<String>());
+        }
+        this.outputVariableDependencies.get(node).addAll(variables);
+    }
+
+    private void addOutputVariableDependency(Node node, String variable) {
+        if (variable == null) {
+            addInputVariableDependencies(node, Collections.emptySet());
+            return;
+        }
+        if (!this.outputVariableDependencies.keySet().contains(node)) {
+            this.outputVariableDependencies.put(node, new TreeSet<String>());
+        }
+        this.outputVariableDependencies.get(node).add(variable);
+    }
+
+
+
+    VariableDependenciesVisitor(RumbleRuntimeConfiguration rumbleRuntimeConfiguration) {
+        this.outputVariableDependencies = new HashMap<>();
+        this.inputVariableDependencies = new HashMap<>();
+        this.rumbleRuntimeConfiguration = rumbleRuntimeConfiguration;
+    }
+
+    @Override
+    protected Void defaultAction(Node node, Void argument) {
+        for (Node child : node.getChildren()) {
+            visit(child, null);
+            addInputVariableDependencies(node, this.inputVariableDependencies.get(child));
+        }
+        if (!this.inputVariableDependencies.containsKey(node)) {
```

Comment:
If I understand correctly, this if block is only triggered when the current node has no children. This logic relies on the implementation of the addInputVariableDependencies() function. As such, I feel it is prone to future bugs.

Reply:
DONE, I added getters.